fetch/fetch.go

```diff
@@ -0,0 +1,526 @@
+		f.handleHashError(batch.ID, err)
+	}
+
+	bytes, err := types.InterfaceToBytes(batch)
+	if err != nil {
+		f.handleHashError(batch.ID, err)
+	}
+	// try sending batch to some random peer
+	retries := 0
+	for {
+		// get random peer
+		p := f.net.GetRandomPeer()
+		err := f.net.SendRequest(fetch, bytes, p, f.receiveResponse, timeoutFunc)
+		// if call succeeded, continue to other Requests
+		if err == nil {
+			break
+		}
+		//todo: mark number of fails per peer to make it low priority
+		retries++
+		if retries > f.cfg.MaxRetiresForPeer {
+			f.handleHashError(batch.ID, ErrCouldNotSend(fmt.Errorf("could not send message")))
+			break
+		}
+		f.log.Error("could not send message to peer %v , retrying, retries : %v", p, retries)
+	}
+}
+
+// handleHashError is called when an error occurred processing batches of the following hashes
+func (f *Fetch) handleHashError(batchHash types.Hash32, err error) {
+	f.log.Error("cannot fetch message %v", err)
+	f.activeBatchM.RLock()
+	batch, ok := f.activeBatches[batchHash]
+	if !ok {
+		f.activeBatchM.RUnlock()
+		f.log.Error("batch invalidated twice %v", batchHash)
+	}
+	f.activeBatchM.RUnlock()
+	f.activeReqM.Lock()
+	for _, h := range batch.Requests {
+		for _, callback := range f.activeRequests[h.Hash] {
+			callback.returnChan <- HashDataPromiseResult{
+				Err:  err,
+				Hash: h.Hash,
+			}
+		}
+		delete(f.activeRequests, h.Hash)
+	}
+	f.activeReqM.Unlock()
+
+	f.activeBatchM.Lock()
+	delete(f.activeBatches, batchHash)
+	f.activeBatchM.Unlock()
+}
+
+// HashDataPromiseResult is the result strict when requesting data corresponding to the Hash
+type HashDataPromiseResult struct {
+	Err  error
+	Hash types.Hash32
+}
+
+// GetAllHashes gets a list of hashes and validates them with the dataCallback function
+// it will return error if any of the hashes were not received.
+func (f *Fetch) GetAllHashes(hashes []types.Hash32, hint Hint, hashValidationFunction HandleDataCallback, validateAndSubmit bool) error {
+	hashWaiting := make([]chan HashDataPromiseResult, len(hashes))
+
+	for _, id := range hashes {
+		callBackChan := f.GetHash(id, hint, hashValidationFunction, validateAndSubmit)
+		hashWaiting = append(hashWaiting, callBackChan)
+	}
+
+	// wait for all block fetch Requests to end
+	var retErr error
+	for _, hashRes := range hashWaiting {
+		res := <-hashRes
+		if res.Err != nil {
+			f.log.Error("%v hash $v", res.Err, res.Hash)
+			retErr = res.Err
+		}
+	}
+
+	return retErr
+}
+
+// GetHash is the regular buffered call to get a specific hash, using provided hash, h as hint the receiving end will
+// know where to look for the hash, if function fails to get a response - failCallback will be called, if a response was
+// received dataCallback will be called. caller can choose whether to validate that the received bytes match the has by setting
+// validateHash to true
+func (f *Fetch) GetHash(hash types.Hash32, h Hint, dataCallback HandleDataCallback, validateAndSubmit bool) chan HashDataPromiseResult {
```

Comment:
you are correct, when designing the feature I thought that all of the data handling can be made in an async way, and that we could aggregate callbacks, when starting to implement the actual fetcher I've noticed that sometimes we do need to verify result synchronously so I had to add the channel. Do you think it will be clearer to separate the functions? the attempt to have a dataCallback was made to standardise how we handle data received. if you think it's best to simplify it more I could make this adjustment

Reply:
I think that:
1. `GetHash` should return a simple `([]byte, err)` response.
1. `GetHashes` (instead of `GetAllHashes`) should return a chan of `(hash, []byte, err)`. Also see my comment here https://github.com/spacemeshos/go-spacemesh/pull/2128#discussion_r513971141.